core-bundle/tests/Command/MigrateCommandTest.php

```diff
@@ -0,0 +1,217 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Tests\Command;
+
+use Contao\CoreBundle\Command\MigrateCommand;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\CoreBundle\Migration\MigrationResult;
+use Contao\CoreBundle\Migration\Migrations;
+use Contao\CoreBundle\Tests\TestCase;
+use Contao\InstallationBundle\Database\Installer;
+use PHPUnit\Framework\MockObject\MockObject;
+use Symfony\Component\Config\FileLocator;
+use Symfony\Component\Console\Tester\CommandTester;
+
+class MigrateCommandTest extends TestCase
+{
+    public function testExecutesWithoutPendingMigrations(): void
+    {
+        $command = $this->getCommand();
+        $tester = new CommandTester($command);
+        $code = $tester->execute([]);
+        $display = $tester->getDisplay();
+
+        $this->assertSame(0, $code);
+        $this->assertRegExp('/All migrations completed/', $display);
+    }
+
+    public function testExecutesPendingMigrations(): void
+    {
+        $command = $this->getCommand(
+            [['Migration 1', 'Migration 2']],
+            [[new MigrationResult(true, 'Result 1'), new MigrationResult(true, 'Result 2')]]
+        );
+
+        $tester = new CommandTester($command);
+        $tester->setInputs(['y']);
+
+        $code = $tester->execute([]);
+        $display = $tester->getDisplay();
+
+        $this->assertSame(0, $code);
+        $this->assertRegExp('/Migration 1/', $display);
+        $this->assertRegExp('/Migration 2/', $display);
+        $this->assertRegExp('/Result 1/', $display);
+        $this->assertRegExp('/Result 2/', $display);
+        $this->assertRegExp('/All migrations completed/', $display);
+    }
+
+    public function testExecutesRunOnceFiles(): void
+    {
+        $runOnceFile = $this->getFixturesDir().'/runonceFile.php';
+
+        file_put_contents($runOnceFile, '<?php $GLOBALS["test_'.__CLASS__.'"] = "executed";');
+
+        $command = $this->getCommand([], [], [[$runOnceFile]]);
+
+        $tester = new CommandTester($command);
+        $tester->setInputs(['y']);
+
+        $code = $tester->execute([]);
+        $display = $tester->getDisplay();
+
+        $this->assertSame('executed', $GLOBALS['test_'.__CLASS__]);
+
+        unset($GLOBALS['test_'.__CLASS__]);
+
+        $this->assertSame(0, $code);
+        $this->assertRegExp('/runonceFile.php/', $display);
+        $this->assertRegExp('/All migrations completed/', $display);
+    }
+
+    public function testExecutesSchemaDiff(): void
+    {
+        $installer = $this->createMock(Installer::class);
```

Comment:
Either the `contao/installation-bundle` is currently a hard dependency of the core-bundle and then the argument doesn't need to be optional in the `Contao\CoreBundle\Command\MigrateCommand` service or this test should be skipped if `!class_exists(Installer::class)`. Or did I miss anything here?

Reply:
The command checks if the installer was provided in https://github.com/contao/contao/pull/709/files#diff-b878e4d8368ed6f024ba9eeafc45f150R188
So the schema diff only works with the installation bundle, but the migrations can still be executed without it.

Reply:
I should probably improve the error message to make it clear that the installation-bundle has to be installed for the schema diff to work.

Reply:
Which isn't really what we want because it should be core functionality. But I'm okay with that for now. As we want to get rid of the installation bundle one day anway. So let's go step by step here :)

Reply:
39bc97273ac36c9a7e4b576765b48acc45388cdf